### DISCUSSION TITLE
Need of different modes to run Play Framework

### DISCUSSION THREAD

**Discussion by janani-reddy9** [playframework/playframework#12283] [General]
I see that we have 3 modes to run the application, Mode.Prod, Mode.Dev, Mode.Test
I am wondering on the use case/purpose, Why do we need to have different modes to run Play or any application?

**Comment by slisaasquatch** [comment]
One use case is that Play's `Environment` maps to Guice's `Stage`, and [it affects whether singletons are eager by default](https://github.com/google/guice/wiki/Scopes).

**Comment by Squeng** [comment]
There are various reasons/use cases, another one is https://www.playframework.com/documentation/3.0.x/PlayConsole#Development-mode for instance.

**Comment by mkurz** [comment]
Like mentioned by others here, there are various reasons why it's useful to know if an application is running in development, test or production:
For example: If you run in dev mode, developing an application locally on your machine, you probably want to see as much detailed stack traces as possible in the error pages in the browser, but in in production you want to hide that from the user. So Play decides based on the mode it's running in, how much information it provides to users. Maybe you have use cases for your development flow so that you want to provide certain functionally / data only in dev mode while developing locally, but not in other modes; you could do that by checking `isDev` on your environment. I have seen projects where they display a big fat red banner on top of their internal admin pages that tell users they are currently looking at the live app in production mode, so they don't by mistake think they are in their local application they are currently working on to not by mistake delete stuff by trying out stuff in the wrong UI etc.
Maybe when developing locally you want to avoid using any caches, so you can bypass them again by checking `isDev`, etc. etc. etc.

There are many use cases where it can be usueful that an app is able to check it's current environment it is running in.
